File: sdk/unity/Assets/Picovoice/Picovoice.cs
﻿//
// Copyright 2021 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
// file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using UnityEngine;
using UnityEngine.Networking;

namespace Pv.Unity
{    
    public class Picovoice : IDisposable
    {        
        private Porcupine _porcupine;
        private Rhino _rhino;
        private Action _wakeWordCallback;
        private Action<Inference> _inferenceCallback;
        
        private bool _isWakeWordDetected;

        /// <summary>
        /// Gets the required number of audio samples per frame.
        /// </summary>
        /// <returns>Required frame length.</returns>
        public int FrameLength { get; private set; }

        /// <summary>
        /// Get the audio sample rate required by Picovoice
        /// </summary>
        /// <returns>Required sample rate.</returns>
        public int SampleRate { get; private set; }

        /// <summary>
        /// Gets the version number of the Picovoice platform
        /// </summary>
        /// <returns>Version of Picovoice</returns>
        public string Version => "1.1.0";

        /// <summary>
        /// Get the version of the Porcupine library
        /// </summary>
        /// <returns>Porcupine library version</returns>
        public string PorcupineVersion { get; private set; }

        /// <summary>
        /// Get the version of the Rhino library
        /// </summary>
        /// <returns>Rhino library version</returns>
        public string RhinoVersion { get; private set; }

        /// <summary>
        /// Gets the current Rhino context information.
        /// </summary>
        /// <returns>Context information</returns>
        public string ContextInfo { get; private set; }

        /// <summary>
        /// Picovoice constructor
        /// </summary>
        /// <param name="keywordPath">Absolute path to Porcupine's keyword model file.</param>
        /// <param name="wakeWordCallback">
        /// User-defined callback invoked upon detection of the wake phrase. 
        /// The callback accepts no input arguments.
        /// </param>
        /// <param name="contextPath">
        /// Absolute path to file containing context parameters. A context represents the set of
        /// expressions(spoken commands), intents, and intent arguments(slots) within a domain of interest.
        /// </param>
        /// <param name="inferenceCallback">
        /// User-defined callback invoked upon completion of intent inference. The callback
        /// accepts a single input argument of type `Map<String, dynamic>` that is populated with the following items:
        /// (1) IsUnderstood: whether Rhino understood what it heard based on the context
        /// (2) Intent: if isUnderstood, name of intent that were inferred
        /// (3) Slots: if isUnderstood, dictionary of slot keys and values that were inferred
        /// </param>
        /// <param name="porcupineModelPath">Absolute path to the file containing Porcupine's model parameters.</param>
        /// <param name="porcupineSensitivity">
        /// Wake word detection sensitivity. It should be a number within [0, 1]. A higher
        /// sensitivity results in fewer misses at the cost of increasing the false alarm rate.
        /// </param>
        /// <param name="rhinoModelPath">Absolute path to the file containing Rhino's model parameters.</param>
        /// <param name="rhinoSensitivity">
        /// Inference sensitivity. It should be a number within [0, 1]. A higher sensitivity value
        /// results in fewer misses at the cost of(potentially) increasing the erroneous inference rate.
        /// </returns>
        public static Picovoice Create(string keywordPath, Action wakeWordCallback,
                                       string contextPath, Action<Inference> inferenceCallback,
                                       string porcupineModelPath = null, float porcupineSensitivity = 0.5f,
                                       string rhinoModelPath = null, float rhinoSensitivity = 0.5f)
        {
            Porcupine porcupine = Porcupine.Create(keywordPaths: new List<string> { keywordPath },
                                                   modelPath: porcupineModelPath,
                                                   sensitivities: new List<float> { porcupineSensitivity });

            Rhino rhino = Rhino.Create(contextPath: contextPath,
                                       modelPath: rhinoModelPath,
                                       sensitivity: rhinoSensitivity);

            if (porcupine.FrameLength != rhino.FrameLength)
            {
                throw new ArgumentException("Porcupine and Rhino frame lengths are different");
            }

            if (porcupine.SampleRate != rhino.SampleRate) 
            {
                throw new ArgumentException("Porcupine and Rhino sample rate are different");
            }

            return new Picovoice(porcupine, wakeWordCallback, rhino, inferenceCallback);
        }

        // private constructor
        private Picovoice(Porcupine porcupine, Action wakeWordCallback, Rhino rhino, Action<Inference> inferenceCallback)
        {
            _porcupine = porcupine;
            _wakeWordCallback = wakeWordCallback;
            _rhino = rhino;
            _inferenceCallback = inferenceCallback;

            FrameLength = porcupine.FrameLength;
            SampleRate = porcupine.SampleRate;
            PorcupineVersion = porcupine.Version;
            RhinoVersion = rhino.Version;
            ContextInfo = rhino.ContextInfo;
        }

        /// <summary>
        /// Processes a frame of the incoming audio stream. Upon detection of wake word and completion of follow-on command
        /// inference invokes user-defined callbacks.
        /// </summary>
        /// <param name="pcm">
        /// A frame of audio samples. The number of samples per frame can be found by calling `.FrameLength`. 
        /// The incoming audio needs to have a sample rate equal to `.SampleRate` and be 16-bit linearly-encoded. 
        /// Picovoice operates on single-channel audio.
        /// </param>
        public void Process(short[] pcm)
        {
            if (pcm.Length != FrameLength)
            {
                throw new ArgumentException(string.Format("Input audio frame size ({0}) was not the size specified by Picovoice engine ({1}). ", pcm.Length, FrameLength) +
                    "Use picovoice.FrameLength to get the correct size.");
            }

            if (_porcupine == null || _rhino == null) 
            {
                throw new ObjectDisposedException("picovoice", "Cannot process frame - resources have been released.");
            }

            if (!_isWakeWordDetected)
            {
                int keywordIndex = _porcupine.Process(pcm);
                if (keywordIndex >= 0)
                {
                    _isWakeWordDetected = true;
                    _wakeWordCallback.Invoke();
                }
            }
            else 
            {
                bool isFinalized = _rhino.Process(pcm);
                if (isFinalized) 
                {
                    _isWakeWordDetected = false;
                    _inferenceCallback.Invoke(_rhino.GetInference());
                }
            }
        }

        /// <summary>
        /// Frees memory that was allocated for Picovoice
        /// </summary>
        public void Dispose()
        {
            if (_porcupine != null) 
            {
                _porcupine.Dispose();
                _porcupine = null;
            }

            if (_rhino != null) 
            {
                _rhino.Dispose();
                _rhino = null;
            }
        }

        ~Picovoice()
        {
            Dispose();
        }
    }
}

File: sdk/go/picovoice.go
// Copyright 2021 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is
// located in the "LICENSE" file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.
//

// Go binding for Picovoice end-to-end platform. Picovoice enables building voice experiences similar to Alexa but
// runs entirely on-device (offline).

// Picovoice detects utterances of a customizable wake word (phrase) within an incoming stream of audio in real-time.
// After detection of wake word, it begins to infer the user's intent from the follow-on spoken command. Upon detection
// of wake word and completion of voice command, it invokes user-provided callbacks to signal these events.

// Picovoice processes incoming audio in consecutive frames. The number of samples per frame is
// `FrameLength`. The incoming audio needs to have a sample rate equal to `SampleRate` and be 16-bit
// linearly-encoded. Picovoice operates on single-channel audio. It uses Porcupine wake word engine for wake word
// detection and Rhino Speech-to-Intent engine for intent inference.

package picovoice

import (
	"fmt"
	"os"

	ppn "github.com/Picovoice/porcupine/binding/go"
	rhn "github.com/Picovoice/rhino/binding/go"
)

// PvStatus descibes error codes returned from native code
type PvStatus int

const (
	SUCCESS          PvStatus = 0
	OUT_OF_MEMORY    PvStatus = 1
	IO_ERROR         PvStatus = 2
	INVALID_ARGUMENT PvStatus = 3
	STOP_ITERATION   PvStatus = 4
	KEY_ERROR        PvStatus = 5
	INVALID_STATE    PvStatus = 6
)

func pvStatusToString(status PvStatus) string {
	switch status {
	case SUCCESS:
		return "SUCCESS"
	case OUT_OF_MEMORY:
		return "OUT_OF_MEMORY"
	case IO_ERROR:
		return "IO_ERROR"
	case INVALID_ARGUMENT:
		return "INVALID_ARGUMENT"
	case STOP_ITERATION:
		return "STOP_ITERATION"
	case KEY_ERROR:
		return "KEY_ERROR"
	case INVALID_STATE:
		return "INVALID_STATE"
	default:
		return fmt.Sprintf("Unknown error code: %d", status)
	}
}

// Callback for when a wake word has been detected
type WakeWordCallbackType func()

// Callback for when Rhino has made an inference
type InferenceCallbackType func(rhn.RhinoInference)

// Picovoice struct
type Picovoice struct {
	// instance of porcupine
	porcupine ppn.Porcupine

	// instance of rhino
	rhino rhn.Rhino

	// only true after init and before delete
	initialized bool

	// true after Porcupine detected wake word
	wakeWordDetected bool

	// Path to Porcupine keyword file (.ppn)
	KeywordPath string

	// Function to be called once the wake word has been detected
	WakeWordCallback WakeWordCallbackType

	// Path to Rhino context file (.rhn)
	ContextPath string

	// Function to be called once Rhino has an inference ready
	InferenceCallback InferenceCallbackType

	// Path to Porcupine model file (.pv)
	PorcupineModelPath string

	// Sensitivity value for detecting keyword. The value should be a number within [0, 1]. A
	// higher sensitivity results in fewer misses at the cost of increasing the false alarm rate.
	PorcupineSensitivity float32

	// Path to Rhino model file (.pv)
	RhinoModelPath string

	// Inference sensitivity. A higher sensitivity value results in
	// fewer misses at the cost of (potentially) increasing the erroneous inference rate.
	// Sensitivity should be a floating-point number within 0 and 1.
	RhinoSensitivity float32

	// Once initialized, stores the source of the Rhino context in YAML format. Shows the list of intents,
	// which expressions map to those intents, as well as slots and their possible values.
	ContextInfo string
}

// Returns a Picovoice stuct with default parameters
func NewPicovoice(keywordPath string,
	wakewordCallback WakeWordCallbackType,
	contextPath string,
	inferenceCallback InferenceCallbackType) Picovoice {
	return Picovoice{
		KeywordPath:       keywordPath,
		WakeWordCallback:  wakewordCallback,
		ContextPath:       contextPath,
		InferenceCallback: inferenceCallback,

		PorcupineSensitivity: 0.5,
		RhinoSensitivity:     0.5,
	}
}

var (
	// Required number of audio samples per frame.
	FrameLength = ppn.FrameLength

	// Required sample rate of input audio
	SampleRate = ppn.SampleRate

	// Version of Porcupine being used
	PorcupineVersion = ppn.Version

	// Version of Rhino being used
	RhinoVersion = rhn.Version

	// Picovoice version
	Version = fmt.Sprintf("1.1.0 (Porcupine v%s) (Rhino v%s)", PorcupineVersion, RhinoVersion)
)

// Init function for Picovoice. Must be called before attempting process.
func (picovoice *Picovoice) Init() error {

	if picovoice.KeywordPath == "" {
		return fmt.Errorf("%s: No valid keyword was provided.", pvStatusToString(INVALID_ARGUMENT))
	}

	if _, err := os.Stat(picovoice.KeywordPath); os.IsNotExist(err) {
		return fmt.Errorf("%s: Keyword file file could not be found at %s", pvStatusToString(INVALID_ARGUMENT), picovoice.KeywordPath)
	}

	if picovoice.ContextPath == "" {
		return fmt.Errorf("%s: No valid context was provided.", pvStatusToString(INVALID_ARGUMENT))
	}

	if _, err := os.Stat(picovoice.ContextPath); os.IsNotExist(err) {
		return fmt.Errorf("%s: Context file could not be found at %s", pvStatusToString(INVALID_ARGUMENT), picovoice.ContextPath)
	}

	if picovoice.InferenceCallback == nil {
		return fmt.Errorf("%s: No InferenceCallback was provided.", pvStatusToString(INVALID_ARGUMENT))
	}

	if ppn.SampleRate != rhn.SampleRate {
		return fmt.Errorf("%s: Pocupine sample rate (%d) was differenct than Rhino sample rate (%d)",
			pvStatusToString(INVALID_ARGUMENT),
			ppn.SampleRate,
			rhn.SampleRate)
	}

	if ppn.FrameLength != rhn.FrameLength {
		return fmt.Errorf("%s: Pocupine frame length (%d) was differenct than Rhino frame length (%d)",
			pvStatusToString(INVALID_ARGUMENT),
			ppn.FrameLength,
			rhn.FrameLength)
	}

	picovoice.porcupine = ppn.Porcupine{
		ModelPath:     picovoice.PorcupineModelPath,
		KeywordPaths:  []string{picovoice.KeywordPath},
		Sensitivities: []float32{0.5},
	}
	err := picovoice.porcupine.Init()
	if err != nil {
		return err
	}

	picovoice.rhino = rhn.Rhino{
		ModelPath:   picovoice.RhinoModelPath,
		ContextPath: picovoice.ContextPath,
		Sensitivity: picovoice.RhinoSensitivity,
	}
	err = picovoice.rhino.Init()
	if err != nil {
		return err
	}
	picovoice.ContextInfo = picovoice.rhino.ContextInfo
	picovoice.initialized = true
	return nil
}

// Releases resouces aquired by Picovoice
func (picovoice *Picovoice) Delete() error {
	porcupineErr := picovoice.porcupine.Delete()
	rhinoErr := picovoice.rhino.Delete()

	if porcupineErr != nil {
		return porcupineErr
	}
	if rhinoErr != nil {
		return rhinoErr
	}

	picovoice.initialized = false
	return nil
}

// Process a frame of pcm audio with the Picovoice platform.
// Invokes user-defined callbacks upon detection of wake word and completion of follow-on command inference
func (picovoice *Picovoice) Process(pcm []int16) error {
	if !picovoice.initialized {
		return fmt.Errorf("Picovoice could not process because it has either not been initialized or has been deleted.")
	}

	if len(pcm) != FrameLength {
		return fmt.Errorf("Input data frame size (%d) does not match required size of %d", len(pcm), FrameLength)
	}

	if !picovoice.wakeWordDetected {
		keywordIndex, err := picovoice.porcupine.Process(pcm)
		if err != nil {
			return err
		}

		if keywordIndex == 0 {
			picovoice.wakeWordDetected = true
			if picovoice.WakeWordCallback != nil {
				picovoice.WakeWordCallback()
			}
		}
	} else {
		isFinalized, err := picovoice.rhino.Process(pcm)
		if err != nil {
			return err
		}
		if isFinalized {
			picovoice.wakeWordDetected = false
			inference, err := picovoice.rhino.GetInference()
			if err != nil {
				return err
			}

			picovoice.InferenceCallback(inference)
		}
	}
	return nil
}
